File: CSDS/B/CSDS-B-B2/8.ProjectCode/frontend/src/lib/devices.ts
export interface Device {
  id: number;
  owner_id: number;
  owner_name: string | null;
  kind: "phone" | "esp32" | "simulator" | "replay";
  name: string;
  api_key_prefix: string;
  hardware: string | null;
  firmware: string | null;
  fixed_lat: number | null;
  fixed_lon: number | null;
  config: Record<string, unknown>;
  is_active: boolean;
  created_at: string;
  last_seen_at: string | null;
  last_sync_at: string | null;
  readings_count: number;
  online: boolean;
}

export interface DeviceWithKey {
  device: Device;
  api_key: string;
}

export interface SyncBatch {
  id: number;
  ts: string;
  received: number;
  accepted: number;
  duplicates: number;
  rejected: number;
  oldest_reading: string | null;
}

export const KIND_LABEL: Record<Device["kind"], string> = {
  phone: "Phone probe",
  esp32: "ESP32 node",
  simulator: "Simulated node",
  replay: "Sample replay",
};

File: CSDS/B/CSDS-B-B2/8.ProjectCode/backend/app/schemas/devices.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

from .common import ORMModel, UTCDateTime


class DeviceOut(ORMModel):
    id: int
    owner_id: int
    owner_name: str | None = None
    kind: str
    name: str
    api_key_prefix: str
    hardware: str | None
    firmware: str | None
    fixed_lat: float | None
    fixed_lon: float | None
    config: dict
    is_active: bool
    created_at: UTCDateTime
    last_seen_at: UTCDateTime | None
    last_sync_at: UTCDateTime | None
    readings_count: int
    online: bool = False


class DeviceCreate(BaseModel):
    kind: Literal["esp32", "simulator"] = "esp32"
    name: str = Field(min_length=2, max_length=120)
    fixed_lat: float | None = Field(default=None, ge=-90, le=90)
    fixed_lon: float | None = Field(default=None, ge=-180, le=180)
    network_name: str | None = Field(default=None, max_length=80)


class PhonePair(BaseModel):
    name: str | None = Field(default=None, max_length=120)
    hardware: str | None = Field(default=None, max_length=255, description="Browser / phone description")


class DeviceUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=2, max_length=120)
    fixed_lat: float | None = Field(default=None, ge=-90, le=90)
    fixed_lon: float | None = Field(default=None, ge=-180, le=180)
    network_name: str | None = Field(default=None, max_length=80)
    is_active: bool | None = None


class DeviceWithKey(BaseModel):
    device: DeviceOut
    api_key: str = Field(description="Shown once - store it on the device")


class SyncBatchOut(ORMModel):
    id: int
    ts: UTCDateTime
    received: int
    accepted: int
    duplicates: int
    rejected: int
    oldest_reading: UTCDateTime | None
